Added first version of benchmarking tutorial (#152)
micro-ROS/micro-ROS.github.io: _docs/tutorials/advanced/benchmarking/index.md

---
title: Benchmarking tooling Tutorial - Shadow Builder
permalink: /docs/tutorials/advanced/benchmarking/
author: Alexandre Malki (improved by Tomasz Kołcon)
---

> The Trace Framework Abstraction is a module that will be part of the shadow

Reviewer: As a reader, I did not understand the relation between TFA and shadow builder. You should try to explain the idea/concept of the Shadow Builder and its key features in two, three sentences and then introduce TFA and the relation of the Shadow Builder.

builder. This module is in charge of abstracting the underlying tracing tool
used to benchmark.  The tool is still a stiLl a work in progress 
(everything is perfectible). So over time new features and bugs resolving will
be pushed.

> ## Table of Contents

Reviewer: In the first part, there is a large overlap with the README.md of the repo at https://github.com/micro-ROS/benchmarking_shadow-builder. For long-term maintainability, I propose to remove the installation instructions and the architecture description here but point to the README.md in the repo and make this page a real tutorial.

  1            /usr/bin/gcc-7   10        manual mode
  2            /usr/bin/gcc-9   20        manual mode

Press <enter> to keep the current choice[*], or type selection number: 2
```

And repeat those steps for:
 * c++ --> selecting /usr/bin/g++
 * cc --> selecting /usr/bin/gcc
 * gcc --> selecting /usr/bin/gcc-9
 * g++ --> selecting /usr/bin/g++-9

It is possible to check if the right version of gcc and g++ in the command line:

``` shell
gcc -v
g++ -v
cc -v
c++ -v
```

If some depencies were to be missing, please feel free to add them into this
document and then initiate a pull request.

### Compilation 

Before starting, some internal dependencies must be retrieved. To do so, the
script at the root of folder should be called as follow:

``` shell
./prepare_build.sh # retrieves/compiles and installs dependencies
```

Once this step is done, you can delete the ext/dl subdirectory. This will save you a lot
of space

**This might take some time, go grab yourself a coffee**

Then once all dependencies are built, the shadow-builder is ready to be
compiled:

``` shell
mkdir -p build; cd build
cmake ..
make -j4
```

### First run: using the example/example_stupid_main

> To run the the shadow buidler against the examples (located ina examples/example_stupid_main),

Reviewer: ```suggestion
To run the the shadow buidler against the examples (located in `examples/example_stupid_main`),
```

> the following steps to perform are:

Reviewer: ```suggestion
perform the following steps:
```

```shell
cd build # previously created during the compilation step.
./shadow-program -s ../res/sb-res/bcf.xml -t ../res/tfa-res/tfa.xmlb
```

This command will create an instrumented code that will be generated in the
folder /tmp/output/test_DATE. 

Once in the folder, the following commands will compile and execute the code:

``` shell
cd /tmp/output/test_DATE/
mkdir -p build; cd build
cmake .. ; make
./simple_stupid_example 
```

The output of this command should be:

``` shell
Monitor var i: 0
Monitor var i: 4294967295
Exe time: 8 sec : 758 ms : 534352 ns
```

> To understand what is going, it is recommanded to go through all configurations

Reviewer: ```suggestion
To understand what is going, it is recommended to go through all configurations
```

file explaination and readmes, the examples code and its readme. 

For more details, refering to all readmes are a good start. The code is
documentated as well and can provide additional information about what is going
on at a very low level.

### Configuration files

The shadow-builder is using XML configuration files to locate the TFA pluging folder 
**currently only one folder can be provided**. By default the plugin folder will

> be located in the folder <src_root>/tfa-plugins/

Reviewer: ```suggestion
be located in the folder `<src_root>/tfa-plugins/`
```

More information regarding the different attributes and node fo the XML

> configuraition can be find here: [res/README.md](res/README.md) .

Reviewer: ```suggestion
configuration can be find here: [res/README.md](res/README.md) .
```

### Plugins

> Plugins are the "Responder" to _Tags_ that are found in the code. This _Tags_

Reviewer: ```suggestion
Plugins are the "responders" to _Tags_ that are found in the code. Such a _Tag_
```

> are used to tell the shadow-buidler and more specifically its plugins that a

Reviewer: ```suggestion
is used to tell the shadow-builder and more specifically its plugins that a
```

> some benchmarking request are made in the code. 

Reviewer: ```suggestion
benchmarking request is made in the code. 
```

The response is up to the implementation in the plugin. For more details, please

> have a look here [tfa-plugins/README.md](tfa-plugins/README.md) .

Reviewer: Local link will not work.

### Unit tests

> Unit tests are available in the folder <module>/tests/xxx_\<xx\>.cpp

Reviewer: ```suggestion
Unit tests are available in the folder <module>/tests/xxx_\<xx\>.cpp
```
```suggestion
Unit tests are available in the folder <module>/tests/xxx_\<xx\>.cpp
```

Each test shall be added in the CMakeFile.txt in the module you are testing.

In addition the option -DENABLE_TESTS=ON shall be passed to the cmake to
activate tests creation.

Curently a few are test are performed. In the future, more unit test will
be added.

## Tested Platforms

> The shadow-builder was successfully compiled with ubuntu 18.04 LTS with the following

Reviewer: ```suggestion
The shadow-builder was successfully compiled with Ubuntu 18.04 LTS with the following
```

		return Status::returnStatusOkay();
	} else if (params[0]  == "stop" && params.size() == 2) {
		std::string start = "timer_start_" + params[1]; 
		std::string stop = "timer_stop_" + params[1]; 
		char buf[sizeof(difftime_func) + start.length() +
			stop.length()];

		sprintf(buf, difftime_func, start.c_str(), stop.c_str());
		replacement += "clock_gettime(CLOCK_PROCESS_CPUTIME_ID,
					&timer_stop_" + params[1] + ");\n";
		replacement += string(buf);
		
		return Status::returnStatusOkay();
	}

	return Status::returnStatusError();
}
```

## Register the listener

Once the listener is implemented, then it needs to to be registerd within the
plugin:

```cpp
bool TimeBenchmarking::initializePlugin()
{	
	/* This is a plugin compatibility platform */
	tbp = new TFABenchMarkingPlatform("Linux", "*", "*", "*")

	/** Here register the Timer listener */
	iclVect.emplace_back(static_cast<ITFACommentListener *>(new
					Timer));

	/* This is the infoPlugin that holds the plugin name and the
		platform information */
`	mInfos = new TFAInfoPlugin("Test Plugin", *tbp);

	// Will be explained later how to mock up this part.
	return Status::returnStatusOkay();
}
```

> A protected vector,from the IPlugin, class needs to be appended for each listener this plugin will be

Reviewer: ```suggestion
A protected vector, from the IPlugin, class needs to be appended for each listener this plugin will be
```
